### PR TITLE
Streamflow related indicators

### PR DESCRIPTION
**Pull request by huard**
* **Please check if the PR fulfills these requirements**
- [x] This PR addresses an already opened issue (for bug fixes / features)
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, etc.)
Adds the following indicators for stream flow values: 
- freq_analysis: compute the n-year return period values
- stats: compute statistical indicators (mean, max, min, std, var, count, sum, argmin, argmax)
- doy_qmax, doy_qmin : day of the year of max or min

This PR makes an effort to reduce code duplication by taking advantage of generic functions. Work to auto-generate more indicators has started but it proved to be difficult to get it right. Another branch called `gen-stats` is created to pursue that work. 

* **Does this PR introduce a breaking change?** (Has there been an API change?)
No

### CODE REVIEW

**Comment by tlogan2000** [comment]
@huard j'essaie de regarder ça ce pm ou demain matin

File: tests/test_streamflow.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
+from xclim import streamflow
+import numpy as np
+
+
+def test_base_flow_index(ndq_series):
```
Comment (tlogan2000) [inline tests/test_streamflow.py]: Tests seem somewhat sparse here testing only for metadata additions and not output values

File: xclim/generic.py [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+    if not indexer:
+        selected = da
+    else:
+        key, val = indexer.popitem()
+        time_att = getattr(da.time.dt, key)
+        selected = da.sel(time=time_att.isin(val)).dropna(dim='time')
+
+    return selected
+
+
```
Comment (tlogan2000) [inline xclim/generic.py]: This is where I have spent most of my time reviewing.  I think I have my head wrapped around it and everything seems functional but I found the mixing of resample and seasonal subsetting using the **indexer in a single function somewhat confusing.   

I wonder if it would it be better to have users explicitly run 'select_time' to subset the data for a given season or month before using freq_analysis or other in a separate call.  This is how I would do it my and I suppose I still could as the indexer is optional but it took me sometime do understand that it does both ...

Comment (tlogan2000) [reply xclim/generic.py]: Are there applications where there is a time subset that would combined with any other resample frequency other than YS?

File: xclim/generic.py [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+    xarray.DataArray
+      An array of values with a 1/t probability of exceedance or non-exceedance when mode is high or low respectively.
+
+    """
+    # Apply rolling average
+    if window > 1:
+        da = da.rolling(time=window, center=False).mean()
+
+    # Assign default resampling frequency if not provided
+    freq = freq or default_freq(**indexer)
```
Comment (tlogan2000) [inline xclim/generic.py]: There is a default freq='YS' in the function parameters so default_freq() would never be used?

File: xclim/generic.py [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+
+    # Extract the time series of min or max over the period
+    sel = select_resample_op(da, op=mode, freq=freq, **indexer).dropna(dim='time')
+
```
Comment (tlogan2000) [inline xclim/generic.py]: I assume we drop nans because the scipy fitting function will not accept them, but do we need to check to make sure there is a large enough sample to properly fit the dist in the subsequent fa() call?

**Comment by tlogan2000** [comment]
I have asked @sbiner to take a look a this as well so I will hold off my approval for now.  My relatively few comments aside everything seems OK for a first version of the streamflow functions.  General comment (maybe because I have little background in hydrology) would be that a good set of tutorial material would be beneficial

File: tests/test_streamflow.py (hunk 1 above)
Comment (huard) [reply tests/test_streamflow.py]: There is an output test in test_indices.

File: xclim/generic.py (hunk 2 above)
Comment (huard) [reply xclim/generic.py]: Yes, for DJF, you'd want YS-DEC. 

To clarify: 
- `select_time` is the function that only selects a time period. 
- `select_resample_op` applies an operation on a selected period at a given resampling frequency. 
- `frequency_analysis` is the convenience function wrapping these operations together:
  - rolling average
  - min/max operation
  - period selection
  - resampling
  - `fa` on the resulting array. 

So basically, you could do the array preparation on your own, then call `fa` on the result. Since there are a lot of streamflow indices to compute (for all seasons, min/max, different rolling averages) I thought it was best to provide that kitchen sink function.

File: xclim/generic.py (hunk 3 above)
Comment (tlogan2000) [reply xclim/generic.py]: i.e Should the function def have freq=None ?

Comment (huard) [reply xclim/generic.py]: Unless you pass None. You've put your finger on the tender spot of the PR. Not clear to me what to do here. One option is to set the default frequency based on the period. So for example, it's always 'YS' unless seasons is DJF. The issue with this is that you might want to take the full year, but split the year in July. So, I think we need to expose the `freq` option, the question is how to handle the default. Options: 
- Set freq=None, and then call default_freq to get the default. The problem is that it's not apparent in the signature what the default value is going to be;
- Set freq='YS', and only call default_freq if freq is None;
- Set freq as a positional argument, so users need to pass a value.

Comment (huard) [reply xclim/generic.py]: Now I remember ! But the reason I've chosen this option is that our Indicator class expects the signature to contain freq and its default. If we change that, I need to go back in Indicators and modify the code. Doable, but I was lazy. So let's decide how it should be, and I'll make it work.

Comment (tlogan2000) [reply xclim/generic.py]: With the code as is I think many users would not think to explicitly set freq=None in order to get the appropriate default for the DJF season (despite the fact that you wrote it in the docstring. e.g. I missed it and was looking fairly closely )

Comment (tlogan2000) [reply xclim/generic.py]: > * Set freq=None, and then call default_freq to get the default. The problem is that it's not apparent in the signature what the default value is going to be;

If changes to the Indicator class are not a major headache .. this would maybe be the best option in my mind.  Could the docstring include a more explicit mention of the defaults by season generated by default_freq()

File: xclim/generic.py (hunk 4 above)
Comment (huard) [reply xclim/generic.py]: That's something that should be implemented in the missing method.

File: xclim/generic.py (hunk 3 above)
Comment (huard) [reply xclim/generic.py]: Yes, let's try that.

Comment (huard) [reply xclim/generic.py]: Done.

File: xclim/generic.py (hunk 4 above)
Comment (huard) [reply xclim/generic.py]: Opened an issue to account for period selection in missing function.
